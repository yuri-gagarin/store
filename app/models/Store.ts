import mongoose, { Schema, Document } from "mongoose";
import { IStorePicture } from "./StorePicture";
type IStorePictureRef = mongoose.Types.ObjectId | IStorePicture;

export interface IStore extends Document {
  title: string;
  description: string;
  images: IStorePictureRef[];
  createdAt: Date;
  editedAt?: Date;
}

const StoreSchema: Schema = new Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  images: [
    {
      type: Schema.Types.ObjectId,
      ref: "StorePicture"
    }
  ],
  createdAt: {
    type: Date,
    default: Date.now,
    required: true
  },
  editedAt: {
    type: Date
  }
});
const Store = mongoose.model<IStore>('Store', StoreSchema);
export default Store;