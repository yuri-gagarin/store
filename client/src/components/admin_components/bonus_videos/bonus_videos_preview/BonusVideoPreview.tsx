import React from "react";
import { Button, Embed, Item } from "semantic-ui-react";
// css //
import "./css/bonusVideoPreview.css";
// helpers //
import { ConvertDate, setDefaultImage, trimString } from "../../../helpers/displayHelpers";

interface Props {
  bonusVideo: IBonusVideoData;
}

const BonusVideoPreview: React.FC<Props> = ({ bonusVideo }): JSX.Element => {
  const { description, vimeoURL, youTubeURL, createdAt } = bonusVideo;

  if (youTubeURL) {
    return (
      <div className="youtubeBonusVideo">
        <Embed
          url={bonusVideo.youTubeURL}
        />
        <div className="bonusVideoDesc">
          {description}
        </div>
      </div>
    );
  } else if (vimeoURL) {
    return (
      <div>
        <Embed 
          url={bonusVideo.vimeoURL}
        />
        <div className="bonusVideoDesc">
          {description}
        </div>
      </div>
    );
  } else {
    return (
      <div className="bonusVideoErrorDiv">
        <div className="bonusVideoErrorDivMsg">Can't Resolve Video Link</div>
      </div>
    )
  }
};

export default BonusVideoPreview;