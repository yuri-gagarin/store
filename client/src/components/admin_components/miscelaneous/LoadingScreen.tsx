import React from 'react'
import { Dimmer, Loader, Segment } from 'semantic-ui-react'

const LoaderExampleText: React.FC<{}> = (props): JSX.Element => {
  return (
    <div style={{ height: "100vh" }} id="generalLoadingScreen"> 
      <Segment style={{ height: "100vh" }}>
        <Dimmer active inverted>
          <Loader>Please Wait</Loader>
        </Dimmer>
      </Segment>
    </div>
  )
};

export default LoaderExampleText