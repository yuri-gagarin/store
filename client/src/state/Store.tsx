import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

