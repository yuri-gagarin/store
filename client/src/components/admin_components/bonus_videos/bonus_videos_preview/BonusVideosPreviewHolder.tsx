import React, { useContext, useEffect } from "react";
import { Grid, Item } from "semantic-ui-react";
// routing //
import { withRouter, RouteComponentProps } from "react-router-dom";
// css imports //
import "./css/bonusVideosPreviewHolder.css";
// additional components //
import BonusVideoPreview from "./BonusVideoPreview";
import BonusVideosControls from "./BonusVideoControls";
import LoadingScreen from "../../miscelaneous/LoadingScreen";
// types and interfaces //
import { Store} from "../../../../state/Store";
// api actions //
import { getAllBonusVideos } from "../actions/APIBonusVideoActions";

interface Props extends RouteComponentProps {
  
}

const BonusVideosPreviewHolder: React.FC<Props> = ({ history }): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, loadedBonusVideos } = state.bonusVideoState;

  // lifecycle hooks //
  useEffect(() => {
    let componentLoaded = true;
    if (componentLoaded) {
      getAllBonusVideos(dispatch)
        .then((_) => {
          // handle success //
        })
        .catch((_) => {
          // handle error //
        });
    }
    return () => { componentLoaded = false };
  }, [ dispatch ]);

  // compnent return //
  return (
    loading ? 
    <LoadingScreen />
    :
    <Grid 
      id={"adminBonusVideosPreviewHolder"}
      stackable 
      padded 
      columns={2}
    >
      <Grid.Row>
      <Grid.Column computer={10} tablet={8} mobile={16}>
        <Item.Group>
          {
            loadedBonusVideos.map((video) => {
              return (
                <BonusVideoPreview 
                  key={video._id}
                  bonusVideo={video}
                />
              );
            })
          }
        </Item.Group>
      </Grid.Column>
      <Grid.Column computer={6} tablet={8} mobile={16}>
        <BonusVideosControls totalBonusVideos={loadedBonusVideos.length} />
      </Grid.Column>

      </Grid.Row>
      
    </Grid>
  );
};
// test export without the router //
export { BonusVideosPreviewHolder };
// default export //
export default withRouter(BonusVideosPreviewHolder);