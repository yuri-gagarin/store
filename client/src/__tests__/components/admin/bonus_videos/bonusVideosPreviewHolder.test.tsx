import React from "react";
import { Grid } from "semantic-ui-react"
import { BrowserRouter as Router } from "react-router-dom";
// test dependencies //
import moxios from "moxios";
import { mount, ReactWrapper } from "enzyme"; 
import { act } from "react-dom/test-utils";
// components //
import BonusVideosPreviewHolder from "../../../../components/admin_components/bonus_videos/bonus_videos_preview/BonusVideosPreviewHolder";
import BonusVideoPreview from "../../../../components/admin_components/bonus_videos/bonus_videos_preview/BonusVideoPreview";
import LoadingScreen from "../../../../components/admin_components/miscelaneous/LoadingScreen";
// state //
import { TestStateProvider } from "../../../../state/Store";
// helpers //
import { generateCleanState } from "../../../../test_helpers/miscHelpers";
import BonusVideossControls from "../../../../components/admin_components/bonus_videos/bonus_videos_preview/BonusVideoControls";

describe("BonusVideoPreviewHolder Component render tests", () => {
  // TEST StprePreviewHolder in its loading state //
  describe("BonusVideoPreviewHolder in 'loading' state", () => {
    let wrapper: ReactWrapper;

    beforeAll(() => {
      const mockState = generateCleanState();
      mockState.bonusVideoState.loading = true;
      wrapper = mount(
        <TestStateProvider mockState={mockState}>
          <Router keyLength={0} >
            <BonusVideosPreviewHolder />
          </Router>
        </TestStateProvider>
      );
    });
    it("Should correctly render", () => {
      const component = wrapper.find(BonusVideosPreviewHolder)
      expect(component).toMatchSnapshot();
    });
    it("Should have a LoadingScreen while 'loading == true'", () => {
      const loadingScreen = wrapper.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(1);
    });
    it("Should not render the '#adminBonusVideoPreviewHolder'", () => {
      const adminBonusVideoPreview = wrapper.find(Grid);
      expect(adminBonusVideoPreview.length).toEqual(0);
    }); 
    it("Should not render any BonusVideoPreview Components", () => {
      const bonus_videoPreviewComponents = wrapper.find(BonusVideoPreview);
      expect(bonus_videoPreviewComponents.length).toEqual(0);
    });

  });
  // END TEST BonusVideoPreviewHolder in its loading state //
  // TEST BonusVideoPreviewHolder in its loaded state //
  describe("BonusVideoPreview in 'loaded' state", () => {
    let wrapper: ReactWrapper; const bonusVideos: IBonusVideoData[] = [];
    beforeAll( async () => {
      // mock data
      const vidData: IBonusVideoData = {
        _id: "1111",
        youTubeURL: "url",
        vimeoURL: "url",
        description: "description",
        createdAt: "now"
      };
      bonusVideos.push(vidData);
      // moxios and component mount //
      moxios.install();
      const mockState = generateCleanState();
      wrapper = mount(
         <TestStateProvider mockState={mockState}>
           <Router>
             <BonusVideosPreviewHolder />
           </Router>
         </TestStateProvider>
      );

      await act( async () => {
        await moxios.stubRequest("/api/bonus_videos", {
          status: 200,
          response: {
            responseMsg: "All ok",
            bonusVideos: bonusVideos
          }
        });
      });
      wrapper.update();
    });
    it("Should correctly render", () => {
      const comp = wrapper.find(BonusVideosPreviewHolder);
      expect(comp).toMatchSnapshot();
    });
    it("Should NOT have a LoadingScreen while 'loading == false'", () => {
      const loadingScreen = wrapper.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(0);
    });
    it("Should render the '#adminBonusVideoPreviewHolder'", () => {
      const adminBonusVideoPreview = wrapper.find(Grid);
      expect(adminBonusVideoPreview.length).toEqual(1);
    });
    it("Should render the 'BonusVideosControls' component", () => {
      const bonusVidControlsComp = wrapper.find(BonusVideossControls);
      expect(bonusVidControlsComp.length).toEqual(1);
    });
    it(`Should render a correct (${bonusVideos.length}) number of BonusVideoPreview Components`, () => {
      const bonus_videoPreviewComponents = wrapper.find(BonusVideoPreview);
      expect(bonus_videoPreviewComponents.length).toEqual(bonusVideos.length);
    });
  });
  // END TEST BonusVideoPreviewHolder in its loaded state //
});