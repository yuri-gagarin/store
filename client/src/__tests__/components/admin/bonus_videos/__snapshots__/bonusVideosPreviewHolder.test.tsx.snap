// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BonusVideoPreviewHolder Component render tests BonusVideoPreview in 'loaded' state Should correctly render 1`] = `
<withRouter(BonusVideosPreviewHolder)>
  <BonusVideosPreviewHolder
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {},
        "path": "/",
        "url": "/",
      }
    }
  >
    <Grid
      columns={2}
      id="adminBonusVideosPreviewHolder"
      padded={true}
      stackable={true}
    >
      <div
        className="ui stackable padded two column grid"
        id="adminBonusVideosPreviewHolder"
      >
        <GridRow>
          <div
            className="row"
          >
            <GridColumn
              computer={10}
              mobile={16}
              tablet={8}
            >
              <div
                className="ten wide computer sixteen wide mobile eight wide tablet column"
              >
                <ItemGroup>
                  <div
                    className="ui items"
                  >
                    <BonusVideoPreview
                      bonusVideo={
                        Object {
                          "_id": "1111",
                          "createdAt": "now",
                          "description": "description",
                          "vimeoURL": "url",
                          "youTubeURL": "url",
                        }
                      }
                      key="1111"
                    >
                      <div
                        className="youtubeBonusVideo"
                      >
                        <Embed
                          url="url"
                        >
                          <div
                            className="ui embed"
                            onClick={[Function]}
                          >
                            <Icon
                              as="i"
                              name="video play"
                            >
                              <i
                                aria-hidden="true"
                                className="video play icon"
                                onClick={[Function]}
                              />
                            </Icon>
                          </div>
                        </Embed>
                        <div
                          className="bonusVideoDesc"
                        >
                          description
                        </div>
                      </div>
                    </BonusVideoPreview>
                  </div>
                </ItemGroup>
              </div>
            </GridColumn>
            <GridColumn
              computer={6}
              mobile={16}
              tablet={8}
            >
              <div
                className="six wide computer sixteen wide mobile eight wide tablet column"
              >
                <BonusVideossControls
                  totalBonusVideos={1}
                >
                  <div
                    className="adminBonusVideosControlsHolder"
                  >
                    <div
                      className="adminBonusVideosControlsTitle"
                    >
                      My Videos Controls
                    </div>
                    <div
                      className="adminBonusVideosControls"
                    >
                      <div
                        className="adminBonusVideosCounter"
                      >
                        <div>
                          Total Linked Videos:
                        </div>
                        <div
                          className="adminBonusVideosCount"
                        >
                          1
                        </div>
                      </div>
                    </div>
                  </div>
                </BonusVideossControls>
              </div>
            </GridColumn>
          </div>
        </GridRow>
      </div>
    </Grid>
  </BonusVideosPreviewHolder>
</withRouter(BonusVideosPreviewHolder)>
`;

exports[`BonusVideoPreviewHolder Component render tests BonusVideoPreviewHolder in 'loading' state Should correctly render 1`] = `
<withRouter(BonusVideosPreviewHolder)>
  <BonusVideosPreviewHolder
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {},
        "path": "/",
        "url": "/",
      }
    }
  >
    <LoaderExampleText>
      <div
        id="generalLoadingScreen"
        style={
          Object {
            "height": "100vh",
          }
        }
      >
        <Segment
          style={
            Object {
              "height": "100vh",
            }
          }
        >
          <div
            className="ui segment"
            style={
              Object {
                "height": "100vh",
              }
            }
          >
            <Dimmer
              active={true}
              inverted={true}
            >
              <DimmerInner
                active={true}
                inverted={true}
              >
                <Ref
                  innerRef={
                    Object {
                      "current": <div
                        class="ui active transition visible inverted dimmer"
                        style="display: flex;"
                      >
                        <div
                          class="content"
                        >
                          <div
                            class="ui text loader"
                          >
                            Please Wait
                          </div>
                        </div>
                      </div>,
                    }
                  }
                >
                  <RefFindNode
                    innerRef={
                      Object {
                        "current": <div
                          class="ui active transition visible inverted dimmer"
                          style="display: flex;"
                        >
                          <div
                            class="content"
                          >
                            <div
                              class="ui text loader"
                            >
                              Please Wait
                            </div>
                          </div>
                        </div>,
                      }
                    }
                  >
                    <div
                      className="ui active transition visible inverted dimmer"
                      onClick={[Function]}
                    >
                      <div
                        className="content"
                      >
                        <Loader>
                          <div
                            className="ui text loader"
                          >
                            Please Wait
                          </div>
                        </Loader>
                      </div>
                    </div>
                  </RefFindNode>
                </Ref>
              </DimmerInner>
            </Dimmer>
          </div>
        </Segment>
      </div>
    </LoaderExampleText>
  </BonusVideosPreviewHolder>
</withRouter(BonusVideosPreviewHolder)>
`;
